src/main/java/org/openrewrite/java/testing/mockito/MockitoJUnitToMockitoExtension.java

```diff
@@ -0,0 +1,219 @@
+    public String getDescription() {
+        return "Replaces MockitoJUnit rules with MockitoExtension.";
+    }
+
+    @Override
+    protected TreeVisitor<?, ExecutionContext> getVisitor() {
+        return new MockitoRuleToMockitoExtensionVisitor();
+    }
+
+    public static class MockitoRuleToMockitoExtensionVisitor extends JavaIsoVisitor<ExecutionContext> {
+        private static final String MOCKITO_RULE_FQN = "org.mockito.junit.MockitoRule";
+        private static final String MOCKITO_TEST_RULE_FQN = "org.mockito.junit.MockitoTestRule";
+
+        private static final String MOCKITO_RULE_INVOCATION_KEY = "mockitoRuleInvocation";
+        private static final String MOCKITO_TEST_RULE_INVOCATION_KEY = "mockitoTestRuleInvocation";
+
+        @SneakyThrows
+        @Override
+        public J.ClassDeclaration visitClassDeclaration(J.ClassDeclaration classDecl, ExecutionContext ctx) {
+
+            J.ClassDeclaration cd = super.visitClassDeclaration(classDecl, ctx);
+            Set<J.VariableDeclarations> mockitoFields = FindFields.find(cd, MOCKITO_RULE_FQN);
+            mockitoFields.addAll(FindFields.find(cd, MOCKITO_TEST_RULE_FQN));
+
+            if (!mockitoFields.isEmpty()) {
+                List<Statement> statements = new ArrayList<>(cd.getBody().getStatements());
+                statements.removeAll(mockitoFields);
+                cd = cd.withBody(cd.getBody().withStatements(statements));
+
+                maybeRemoveImport(MOCKITO_RULE_FQN);
+                maybeRemoveImport(MOCKITO_TEST_RULE_FQN);
+
+                maybeRemoveImport("org.junit.Rule");
+                maybeRemoveImport("org.mockito.junit.MockitoJUnit");
+                maybeRemoveImport("org.mockito.quality.Strictness");
+
+                final String extendWithFqn = "org.junit.jupiter.api.extension.ExtendWith";
+                final String mockitoExtensionFqn = "org.mockito.junit.jupiter.MockitoExtension";
+
+                if (classDecl.getBody().getStatements().size() != cd.getBody().getStatements().size()) {
+                    cd = cd.withTemplate(
+                            template("@ExtendWith(MockitoExtension.class)")
+                                    .javaParser(JAVA_PARSER.get())
+                                    .imports(extendWithFqn, mockitoExtensionFqn)
+                                    .build(),
+                            cd.getCoordinates()
+                                    .addAnnotation(Comparator.comparing(
+                                            J.Annotation::getSimpleName,
+                                            new RuleBasedCollator("< ExtendWith")
+                                    ))
+                    );
+
+                    maybeAddImport(extendWithFqn);
+                    maybeAddImport(mockitoExtensionFqn);
+
+                    // @ExtendsWith(MockitoExtension.class) depends on the JUnit5 test annotation to work.
+                    doAfterVisit(new UpdateTestAnnotation());
```

Comment:
Delete these two lines.

It isn't so much that `@ExtendsWith(MockitoExtension.class)` depends on the JUnit5 version of `@Test` as it is that `@ExtendsWith` is itself a JUnit 5 concept. Upgrading just the `@Test` annotation and leaving the rest of the project JUnit4 would obviously not be enough. So to really be sure you'd have to `doAfterVisit()` the whole migration recipe. But that migration recipe is going to _include_ this recipe, so that'd be a mess. Best case makes performance worse. Worst case messes things up some way I don't anticipate.

Reply:
Alright, I'll create a test with an existing annotation, and make sure it isn't added if it already exists.